waiter/src/waiter/process_request.clj

```diff
@@ -190,143 +190,107 @@
+                                        (try
+                                          (when (pos? bytes-read)
+                                            (swap! bytes-streamed-atom + bytes-read)
+                                            (swap! statsd-unreported-bytes-atom + bytes-read))
+                                          (if complete?
+                                            (let [bytes-streamed @bytes-streamed-atom
+                                                  unreported-bytes-to-statsd @statsd-unreported-bytes-atom]
+                                              (log/info "streamed" bytes-streamed "bytes from request input stream")
+                                              (histograms/update! (metrics/service-histogram service-id "request-size") bytes-streamed)
+                                              (statsd/inc! metric-group "request_bytes" unreported-bytes-to-statsd))
+                                            (let [unreported-bytes-to-statsd @statsd-unreported-bytes-atom]
+                                              (when (>= unreported-bytes-to-statsd 1000000)
+                                                (statsd/inc! metric-group "request_bytes" unreported-bytes-to-statsd)
+                                                (swap! statsd-unreported-bytes-atom - unreported-bytes-to-statsd))))
+                                          (catch Throwable throwable
+                                            (log/error throwable "error in reporting request size metrics")))))
+        complete-request-streaming (fn complete-request-streaming [throwable]
+                                     (cid/with-correlation-id
+                                       correlation-id
+                                       (when throwable
+                                         (log/error throwable "unable to stream request bytes, aborting request")
+                                         (async/>!! abort-ch throwable))
+                                       (report-request-size-metrics 0 true)
+                                       (async/close! body-ch)))]
     (try
-      (submit-request-streaming-task executor stream-http-request-fn)
+      (.setReadListener
+        input-stream
+        (reify ReadListener
+          (onAllDataRead [_]
+            (complete-request-streaming nil))
+          (onDataAvailable [_]
+            (try
+              (cid/with-correlation-id
+                correlation-id
+                (log/info "reading available bytes from request input stream")
+                (loop []
+                  (let [bytes-read (read-available-bytes!! streaming-timeout-ms complete-request-streaming input-stream body-ch)]
+                    (when (pos? bytes-read)
+                      (report-request-size-metrics bytes-read false))
+                    (when (and (utils/non-neg? bytes-read) (.isReady input-stream))
```

Comment:
All the examples I've seen have `isReady()` guarding `read()` calls.  Should we check `isReady()` before calling `read()`?

Reply:
The example at https://docs.oracle.com/javaee/7/tutorial/servlets013.htm is using a `do ... while()` loops where the call to `read()` is before the call to `isReady()`